### PR TITLE
changefeedccl: prerequisite changes for `DROP COLUMN`

### PR DESCRIPTION
**Pull request by ajwerner**
This is the first two commits from https://github.com/cockroachdb/cockroach/pull/84563. They are needed to ensure that we don't change the behavior of `DROP COLUMN` when we support it in the declarative schema changer. The issue is that there the protocol is to create a new primary index and swap to it. The column becomes a non-public before the index swap, so the primary index swap is no longer a logical schema change of any kind. With this change, we can now detect that and properly restart as opposed to stop.

Also, the newly added testing uncovers some badness in how we classify some other schema changes, and should generally be useful.

### CODE REVIEW

File: pkg/ccl/changefeedccl/schemafeed/table_event_filter_datadriven_test.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+				now := ts.Clock().Now()
+				targets := parseTargets(t, d.Input)
+				f := schemafeed.New(ctx, cfg, schemafeed.TestingAllEventFilter, targets, now, nil, changefeedbase.CanHandle{
+					MultipleColumnFamilies: true,
+					VirtualColumns:         true,
+				})
+				schemaFeeds[i] = f
+
+				go func() {
+					if err := f.Run(ctx); !assert.NoError(t, err) {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/schemafeed/table_event_filter_datadriven_test.go:116]: i thought it's not safe to require/assert inside go routines like that...?
tests appear to be stuck, I think?

**Review by miretskiy — APPROVED** [review]

File: pkg/ccl/changefeedccl/schemafeed/table_event_filter_datadriven_test.go (hunk 1 above)
Comment (ajwerner) [reply pkg/ccl/changefeedccl/schemafeed/table_event_filter_datadriven_test.go:116]: `assert` is safe, `require` is not.

File: pkg/ccl/changefeedccl/schemafeed/table_event_filter_datadriven_test.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+//
+// - "exec"
+//   Executes the input SQL query.
+//
+// - "create" f=<int>
+//   Creates a schemafeed with the targets specified as the input with the
+//   provided ID.
+//
+// - "pop" f=<int>
+//   Pop all events from the schemafeed with the given ID.
```
[the remark this answers is not in the document]
Comment (ajwerner) [reply pkg/ccl/changefeedccl/schemafeed/table_event_filter_datadriven_test.go:44]: Done

**Review by HonoreDB — APPROVED** [review]
> Reviewed 9 of 10 files at r1, 3 of 5 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84571)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @miretskiy)

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Pull request merged**